api/types/saml_idp_service_provider.go

```diff
@@ -143,3 +144,331 @@ func (s SAMLIdPServiceProviders) Less(i, j int) bool { return s[i].GetName() < s
+	if sortBy.Field == "" {
+		return nil
+	}
+
+	isDesc := sortBy.IsDesc
+	switch sortBy.Field {
+	case ResourceMetadataName:
+		sort.SliceStable(s, func(i, j int) bool {
+			return stringCompare(s[i].GetAppOrServiceProviderName(), s[j].GetAppOrServiceProviderName(), isDesc)
+		})
+	case ResourceSpecDescription:
+		sort.SliceStable(s, func(i, j int) bool {
+			return stringCompare(s[i].GetAppOrServiceProviderDescription(), s[j].GetAppOrServiceProviderDescription(), isDesc)
+		})
+	case ResourceSpecPublicAddr:
+		sort.SliceStable(s, func(i, j int) bool {
+			return stringCompare(s[i].GetAppOrServiceProviderPublicAddr(), s[j].GetAppOrServiceProviderPublicAddr(), isDesc)
+		})
+	default:
+		return trace.NotImplemented("sorting by field %q for resource %q is not supported", sortBy.Field, KindAppAndIdPServiceProvider)
+	}
+
+	return nil
+}
+
+// CheckAndSetDefaults checks and sets default values for any missing fields.
+func (a *AppServerOrSAMLIdPServiceProviderV1) CheckAndSetDefaults() error {
+	if a.IsAppServer() {
+		if err := a.GetAppServer().CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		if err := a.GetSAMLIdPServiceProvider().CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	return nil
+}
+
+func (a *AppServerOrSAMLIdPServiceProviderV1) Expiry() time.Time {
+	if a.IsAppServer() {
+		appServer := a.GetAppServer()
+		return appServer.Metadata.Expiry()
+	} else {
+		sp := a.GetSAMLIdPServiceProvider()
+		return sp.Metadata.Expiry()
+	}
+}
+
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetAllLabels() map[string]string {
+	if a.IsAppServer() {
+		appServer := a.GetAppServer()
+		staticLabels := make(map[string]string)
+		for name, value := range appServer.Metadata.Labels {
+			staticLabels[name] = value
+		}
+
+		var dynamicLabels map[string]CommandLabelV2
+		if appServer.Spec.App != nil {
+			for name, value := range appServer.Spec.App.Metadata.Labels {
+				staticLabels[name] = value
+			}
+
+			dynamicLabels = appServer.Spec.App.Spec.DynamicLabels
+		}
+
+		return CombineLabels(staticLabels, dynamicLabels)
+	} else {
+		sp := a.GetSAMLIdPServiceProvider()
+		return sp.Metadata.Labels
+	}
+}
+
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetLabel(key string) (value string, ok bool) {
+	if a.IsAppServer() {
+		appServer := a.GetAppServer()
+		if cmd, ok := appServer.Spec.App.Spec.DynamicLabels[key]; ok {
+			return cmd.Result, ok
+		}
+
+		v, ok := appServer.Spec.App.Metadata.Labels[key]
+		return v, ok
+	} else {
+		return "", true
+	}
+}
+
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetMetadata() Metadata {
+	if a.IsAppServer() {
+		appServer := a.GetAppServer()
+		return appServer.Metadata
+	} else {
+		sp := a.GetSAMLIdPServiceProvider()
+		return sp.Metadata
+	}
+}
+
+// GetName returns the name of either the App or the SAMLIdPServiceProvider, depending on which one
+// the AppServerOrSAMLIdPServiceProvider holds.
+func (a *AppServerOrSAMLIdPServiceProviderV1) GetName() string {
```

Comment:
what's the difference between this and GetAppOrServiceProviderName?

Reply:
The difference was that `GetName` returned the `Metadata.Name` field while `GetAppOrServiceProviderName` used the `GetName()` methods. I don't think the former is actually necessary, so I removed it and renamed `GetAppOrServiceProviderName` to `GetName`.